File: Chapter 04/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter04/section_4_3_cross_cutting_concerns/before/OrderService.java
package com.grokkingsoftwarearchitecture.chapter04.section_4_3_cross_cutting_concerns.before;

import com.grokkingsoftwarearchitecture.chapter04.shared.LogManager;

/**
 * BUSINESS LOGIC LAYER.
 * ARCHITECTURE WARNING: This class is "welded" to the StaticFileLogger.
 * You cannot test SaveOrder without also executing the static
 * logger logic. This violates the Dependency Inversion Principle.
 */
public class OrderService {
    public void saveOrder(Order order) {
        // HIDDEN DEPENDENCY: This is not visible in the API.
        StaticFileLogger.log("Saving order..." + order.getId());
        LogManager.info(OrderService.class, "(BEFORE_SERVICE) Order saved.");
    }
}

File: Chapter 04/Python/section_4_3_cross_cutting_concerns/before/order_service.py
from .static_file_logger import StaticFileLogger
from shared.log_manager import LogManager

class OrderService:
    """
    ARCHITECTURE WARNING: This class has a hidden dependency.
    It relies on an external global logger, which prevents 
    it from being a "Pure" business logic component.
    """
    def save_order(self, order):
        # 🚨 VIOLATION: Hidden, rigid dependency.
        StaticFileLogger.log("Saving order...")
        LogManager.info("OrderService", "(BEFORE_SERVICE) Order saved.")